### PR TITLE
Cleanup user facing price estimation error

### PR DESCRIPTION
**Pull request by fleupold**
#1823 changed the way we order price estimation errors so that we can better decide whether a failed response is cache worthy (ie we don't want to resubmit a lot of requests for NoLiquidity tokens). 

For instance we now rank a RateLimited (retryable, should not be cached) over a NoLiquidity error (the reason we need priority in the first place is because multiple estimator's responses get merged together in the NativePriceEstimator)

Unintentionally, given the current response propagation, this leads to us showing users e.g. a 429 errors if all estimators return NoLiquidity, but one solver is rate limited (we want to retry that solver next time, but the user should still see "NoLiquidity). Moreover, all uncategorized errors are combined in a "Other" type which yields a 500 to the user.

This PR introduces a differentiation between `SolverInternal` 400 errors (which we want to treat like "no liquidity was found") and `ProtocolInternal` errors (which may indeed be 500 worthy, because something is wrong on our side).

We keep `BadRequest` errors separate (currently some of them are decided by the solvers, some by the protocol) as they indicate the user can change something about their request in order to get quotes to work.

I think this refactoring just scratches the surface (happy to further clean up and differentiate the data models given feedback). Maybe it makes sense to turn PriceEstimationError into an Either of SolverInteranl and ProtocolInternal and have each of them use more verbose subtypes?

I also went through swagger and made sure that the response types for our quote and order placement requests are accurate.

### Test Plan

Unit tests still passing

### CODE REVIEW

**Comment by fleupold** [comment]
https://github.com/cowprotocol/services/pull/new/two_high_level_errors is a preview of how things would like if we separate on the `PriceEstimationError` between `Estimator` and `Protocol` errors.

**Comment by nlordell** [comment]
Another approach would be to move caching to the level of each estimator instead of all estimators. This way, we can keep `NoLiquidity` ranking higher than `RateLimited` (which makes more sense to me from a price estimator consumer point of view) but still retry rate limited price estimates. This also has the added benefit that if, say, 0x returned a `NoLiquidity` error and ParaSwap returned a `RateLimited` error, we would only retry the ParaSwap request and not both.

I think the code is fine as is, but this error handling logic is starting to get unruly, so maybe its time to rethink the abstraction into something that fits a bit better?

Another possible suggestion would be to differentiate "price estimation APIs" from the price estimator that is used for quoting so they can return different errors (instead of "leaking" internal vs external errors to the caller - i.e. the quoting code and API adapter).

**Comment by nlordell** [comment]
> I still think in the old world it was weird that we would potentially tell a user they are rate limited when really just all our price estimators are rate limited (likely from other users using the system).

Agreed. This should have been a 500 internal server error.

**Comment by fleupold** [comment]
Any preference between the enum proposed here and

```rust
#[derive(Error, Debug, Clone)]
pub enum PriceEstimationError {
    #[error("EstimatorPriceEstimationError: {0}")]
    Estimator(#[from] EstimatorPriceEstimationError),
    #[error("ProtocolPriceEstimationError: {0}")]
    Protocol(#[from] ProtocolPriceEstimationError),
}

/// Error that is "on the protocol side" of the price estimation
/// logic (e.g. gas estimator failing, etc.). It's in our power
/// to deal with and fix those errors.
#[derive(Error, Debug)]
pub enum ProtocolPriceEstimationError {
    #[error("token {token:?} is not supported: {reason:}")]
    UnsupportedToken { token: H160, reason: String },

    #[error("Zero Amount")]
    ZeroAmount,

    #[error(transparent)]
    Other(anyhow::Error),
}

/// Error that is "on the external estimator side" of the price estimation
/// logic (e.g. individual rate limits, execution failures, etc.). It's not in
/// our power to deal with and fix those errors.
#[derive(Error, Debug)]
pub enum EstimatorPriceEstimationError {
    #[error("No liquidity")]
    NoLiquidity,

    #[error("Rate limited")]
    RateLimited,

    #[error("Deadline Exceeded")]
    DeadlineExceeded,

    #[error("Unsupported Order Type: {0}")]
    UnsupportedOrderType(String),

    #[error(transparent)]
    Other(anyhow::Error),
}
```

as proposed in https://github.com/cowprotocol/services/compare/two_high_level_errors?expand=1 ?

The one here causes less changes (the other one still needs to update the unit tests)

**Comment by nlordell** [comment]
LGTM (although I still maintain that the fact that there are two errors classes here means that there should be two traits for price estimation - one for estimators and one for the top level price estimation component used in order quoting)

**Comment by fleupold** [comment]
> LGTM

Which of the two?

> here should be two traits for price estimation

💯 agree with this. Unfortunately right now even each individual price estimator relies on logic that can fail for reason in our control. Overall, the quoting logic is quite messy and full with leaky types (the top level compenent called from the API returns a `OrderQuoteError::CalculateQuote::PriceEstimationError`). I'm not sure how much of it I can refactor before the next release (probably not much)

File: crates/shared/src/trade_finding.rs [hunk 1]
```diff
@@ -142,11 +142,16 @@ impl From<PriceEstimationError> for TradeError {
...
             PriceEstimationError::NoLiquidity => Self::NoLiquidity,
-            PriceEstimationError::UnsupportedOrderType => Self::UnsupportedOrderType,
-            PriceEstimationError::DeadlineExceeded => Self::DeadlineExceeded,
+            PriceEstimationError::UnsupportedOrderType(order_type) => {
+                Self::UnsupportedOrderType(order_type)
+            }
+            PriceEstimationError::UnsupportedToken { token, .. } => {
+                Self::UnsupportedOrderType(format!("{token:#x}"))
+            }
+            PriceEstimationError::ZeroAmount => Self::UnsupportedOrderType("zero amount".into()),
```
Comment (MartinquaXD) [inline crates/shared/src/trade_finding.rs:151]: These error conversions seem surprising.
I assumed `UnsupportedOrderType` exclusively means that `buy` or `sell` orders are not supported by that estimator. Was that changed to include anything about an order not being supported (weird amounts, order type, hooks, signatures, send ETH to SC, ...)?
If so maybe something like `UnsupportedOrderConfiguration` might be more fitting?

**Review by MartinquaXD — APPROVED** [review]
I always have a hard time reviewing error conversions but this looks reasonable to me.
This should not be the end game but I think it's a good enough refactor to make it valuable for the next release.

**Review by sunce86 — APPROVED** [review]

**Review by nlordell — APPROVED** [review]

File: crates/shared/src/trade_finding.rs (hunk 1 above)
Comment (fleupold) [reply crates/shared/src/trade_finding.rs:151]: Right, technically for those types of errors we have the `PartialValidationError` type. Technically all queries that reach a price estimator should already be checked for that (ideally the type on `Query.in_amount` would be NonZeroU256).

PriceEstimationError::ZeroAmount is currently checked for in the Baseline solver and in the sanitizing estimator. I don't have a full picture of the control flow to say with confidence that we can never receive these requests there (and could potentially panic if removed)

**Pull request merged**